ftl/python/builder.py

```diff
@@ -0,0 +1,150 @@
+import os
+import subprocess
+import tempfile
+import logging
+import json
+import datetime
+
+from containerregistry.client.v2_2 import append
+from containerregistry.transform.v2_2 import metadata
+
+from ftl.common import builder
+
+_PYTHON_NAMESPACE = 'python-requirements-cache'
+_REQUIREMENTS_TXT = 'requirements.txt'
+_DEFAULT_TTL_WEEKS = 1
+
+
+class Python(builder.JustApp):
+    def __init__(self, ctx):
+        self._overrides = None
+        super(Python, self).__init__(ctx)
+
+    def __enter__(self):
+        """Override."""
+        return self
+
+    def CreatePackageBase(self, base_image, cache, use_cache=True):
+        """Override."""
+        # Figure out if we need to override entrypoint.
+        # Save the overrides for later to avoid adding an extra layer.
+        descriptor = None
+        for f in [_REQUIREMENTS_TXT]:
+            if self._ctx.Contains(f):
+                descriptor = f
+                descriptor_contents = self._ctx.GetFile(f)
+                break
+
+        if not descriptor:
+            logging.info('No package descriptor found. No packages installed.')
+        overrides = metadata.Overrides(
+            creation_time=str(datetime.date.today()) + "T00:00:00Z")
+        # TODO(aaron-prindle) currently if no requirements.txt, this fails here
+        checksum = hashlib.sha256(descriptor_contents).hexdigest()
+        if use_cache:
+            hit = cache.Get(base_image, _PYTHON_NAMESPACE, checksum)
+            if hit:
+                logging.info('Found cached dependency layer for %s' % checksum)
+                # TODO(aaron-prindle) check that cached dep layer is in TTL
+                last_created = _timestamp_to_time(_creation_time(hit))
+                now = datetime.datetime.now()
+                if last_created > now - datetime.timedelta(
+                        seconds=_DEFAULT_TTL_WEEKS):
+                    return hit
+                else:
+                    logging.info('TTL expired for cached image, rebuilding %s'
+                                 % checksum)
+            else:
+                logging.info('No cached dependency layer for %s' % checksum)
+        else:
+            logging.info(
+                'Skipping checking cache for dependency layer %s' % checksum)
+
+        layer, sha = self._gen_package_tar(descriptor, descriptor_contents)
+
+        with append.Layer(
+                base_image, layer, diff_id=sha,
+                overrides=overrides) as dep_image:
+            if use_cache:
+                logging.info('Storing layer %s in cache.', sha)
+                cache.Store(base_image, _PYTHON_NAMESPACE, checksum, dep_image)
+            else:
+                logging.info('Skipping storing layer %s in cache.', sha)
+            return dep_image
+
+    def _gen_package_tar(self, descriptor, descriptor_contents):
+        tmp_app = tempfile.mkdtemp()
+        tmp_venv = tempfile.mkdtemp()
+
+        app_dir = os.path.join(tmp_app, 'app')
```

Comment:
Can app_dir just be tmp_app? I don't think the name is actually important, like it is for venv.